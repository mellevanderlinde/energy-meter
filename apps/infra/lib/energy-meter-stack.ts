import { Duration, RemovalPolicy, Stack, type StackProps } from "aws-cdk-lib";
import { ComparisonOperator } from "aws-cdk-lib/aws-cloudwatch";
import { SnsAction } from "aws-cdk-lib/aws-cloudwatch-actions";
import { AttributeType, BillingMode, Table } from "aws-cdk-lib/aws-dynamodb";
import { PolicyStatement, User } from "aws-cdk-lib/aws-iam";
import { Topic } from "aws-cdk-lib/aws-sns";
import { EmailSubscription } from "aws-cdk-lib/aws-sns-subscriptions";
import type { Construct } from "constructs";

export class EnergyMeterStack extends Stack {
  constructor(
    scope: Construct,
    id: string,
    props: StackProps & { email: string },
  ) {
    super(scope, id, props);

    const appName = "energy-meter";

    const user = new User(this, "User", { userName: appName });

    const table = new Table(this, "Table", {
      tableName: appName,
      billingMode: BillingMode.PAY_PER_REQUEST,
      partitionKey: {
        name: "date",
        type: AttributeType.STRING,
      },
      sortKey: {
        name: "time",
        type: AttributeType.STRING,
      },
      removalPolicy: RemovalPolicy.DESTROY,
    });

    user.addToPolicy(
      new PolicyStatement({
        actions: ["dynamodb:PutItem"],
        resources: [table.tableArn],
      }),
    );

    const topic = new Topic(this, "Topic", {
      topicName: appName,
      enforceSSL: true,
    });

    topic.addSubscription(new EmailSubscription(props.email));

    const noWritesAlarm = table
      .metricConsumedWriteCapacityUnits({
        period: Duration.hours(1),
        statistic: "sum",
      })
      .createAlarm(this, "Alarm", {
        alarmName: appName,
        alarmDescription: "No writes to table",
        evaluationPeriods: 1,
        threshold: 0,
        comparisonOperator: ComparisonOperator.LESS_THAN_OR_EQUAL_TO_THRESHOLD,
      });

    noWritesAlarm.addAlarmAction(new SnsAction(topic));
  }
}
