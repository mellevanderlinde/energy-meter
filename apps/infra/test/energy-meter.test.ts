import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { EnergyMeterStack } from "../lib/energy-meter-stack";

it("should match with snapshot", () => {
  const app = new App();
  const stack = new EnergyMeterStack(app, "Stack", {
    email: "test@example.com",
  });
  const template = Template.fromStack(stack);
  expect(template).toMatchSnapshot();
});
