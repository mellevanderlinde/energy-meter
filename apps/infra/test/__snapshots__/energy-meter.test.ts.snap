// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match with snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Alarm7103F465": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "TopicBFC7AF6E",
          },
        ],
        "AlarmDescription": "No writes to table",
        "AlarmName": "energy-meter",
        "ComparisonOperator": "LessThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "TableName",
            "Value": {
              "Ref": "TableCD117FA1",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ConsumedWriteCapacityUnits",
        "Namespace": "AWS/DynamoDB",
        "Period": 3600,
        "Statistic": "Sum",
        "Threshold": 0,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "TableCD117FA1": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "date",
            "AttributeType": "S",
          },
          {
            "AttributeName": "time",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "date",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "time",
            "KeyType": "RANGE",
          },
        ],
        "TableName": "energy-meter",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "TopicBFC7AF6E": {
      "Properties": {
        "TopicName": "energy-meter",
      },
      "Type": "AWS::SNS::Topic",
    },
    "Topictestexamplecom07E55C1F": {
      "Properties": {
        "Endpoint": "test@example.com",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "TopicBFC7AF6E",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "User00B015A1": {
      "Properties": {
        "UserName": "energy-meter",
      },
      "Type": "AWS::IAM::User",
    },
    "UserDefaultPolicy1F97781E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "dynamodb:PutItem",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "TableCD117FA1",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "UserDefaultPolicy1F97781E",
        "Users": [
          {
            "Ref": "User00B015A1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
